BlobForwarder/index.js

```diff
@@ -42,29 +42,45 @@ module.exports = async function main(context, myBlob) {
     context.log.warn('logs format is invalid');
     return;
   }
-  let compressedPayload;
-  let payloads = generatePayloads(buffer, context);
-  for (const payload of payloads) {
-    try {
-      compressedPayload = await compressData(JSON.stringify(payload));
-      try {
-        await retryMax(httpSend, NR_MAX_RETRIES, NR_RETRY_INTERVAL, [
+
+  let logLines = appendMetaDataToAllLogLines(buffer, context);
+  await compressAndSend(logLines, context);
+};
+
+function compressAndSend(data, context) {
+  return compressData(JSON.stringify(getPayload(data, context)))
+    .then((compressedPayload) => {
+      if (JSON.stringify(compressedPayload).length > NR_MAX_PAYLOAD_SIZE) {
+        let halfwayThrough = Math.floor(data.length / 2);
+
+        let arrayFirstHalf = data.slice(0, halfwayThrough);
+        let arraySecondHalf = data.slice(halfwayThrough, data.length);
+
+        return Promise.all([
+          compressAndSend(arrayFirstHalf, context),
+          compressAndSend(arraySecondHalf, context),
+        ]);
```

Comment:
We need to create another if block here that detects if the `compressedPayload` size is above `NR_MAX_PAYLOAD_SIZE` _AND_ `data.length === 1`. In this case, we can't go recursive any further and we should emit a log.error() message explaining that a message could not be sent because once compressed it was above 1MB, and that we will be discarding it. But we should keep processing the rest of logs. You can use [this](https://github.com/newrelic/newrelic-fluent-bit-output/blob/master/record/record.go#L128-L145) as an example.

Reply:
Nice catch! 🙌